core/src/main/java/org/apache/spark/unsafe/map/BytesToBytesMap.java

```diff
@@ -647,16 +707,24 @@ public boolean putNewKey(
    * Acquire a new page from the memory manager.
    * @return whether there is enough space to allocate the new page.
    */
-  private boolean acquireNewPage() {
-    MemoryBlock newPage = taskMemoryManager.allocatePage(pageSizeBytes);
-    if (newPage == null) {
-      logger.debug("Failed to acquire {} bytes of memory", pageSizeBytes);
+  private boolean acquireNewPage(long required) {
+    try {
+      currentPage = allocatePage(required);
+      dataPages.add(currentPage);
+      Platform.putInt(currentPage.getBaseObject(), currentPage.getBaseOffset(), 0);
+      pageCursor = 4;
+      return true;
+    } catch (OutOfMemoryError e) {
```

Comment:
If we want to catch OOME here, I think that we should do it at a much smaller scope (in the assignment to `currentPage` but not for adding to `dataPages` or modifying the page cursor. Given the risks of catching OOME that I mentioned above, the scope of the catch should be as narrow as possible.